File: Java/src/com/dbc/algorithm/Array/FindFirstAndLastPositionOfElementInSortedArray.java
package com.dbc.algorithm.Array;

public class FindFirstAndLastPositionOfElementInSortedArray {
    public static int binarySearch(int[] nums, int target, boolean lower) {
        int left = 0, right = nums.length - 1, result = nums.length;
        while (left <= right) {
            int mid = (left + right) / 2;
            if (nums[mid] > target || (lower && nums[mid] >= target)) {
                right = mid - 1;
                result = mid;
            } else {
                left = mid + 1;
            }
        }
        return result;
    }

    public static int[] searchRange(int[] nums, int target) {
        int left = binarySearch(nums, target, true);
        int right = binarySearch(nums, target, false) - 1;
        if (left <= right && right < nums.length && nums[left] == target && nums[right] == target) {
            return new int[]{left, right};
        }
        return new int[]{-1, -1};
    }
}

File: Python/array/find_first_and_last_position_of_element_in_sorted_array.py
from typing import List

def binary_search(nums: List[int], target: int, lower: bool) -> int:
    n = len(nums)
    left, right, ans = 0, n - 1, n
    while left <= right:
        index = (left + right) // 2
        mid = nums[index]
        if mid > target or (lower and mid >= target):
            right = index - 1
            ans = index
        else:
            left = index + 1
    
    return ans

def searchRange(nums: List[int], target: int) -> List[int]:
    left = binary_search(nums, target, True)
    right = binary_search(nums, target, False) - 1
    if left <= right and right < len(nums) and nums[left] == target and nums[right] == target:
        return [left, right]
    return[-1, -1]


if __name__ == "__main__":
    print(searchRange(nums = [5,7,7,8,8,10], target = 8))